test/module/shared_model/backend_proto/proto_query_response_factory_test.cpp

```diff
@@ -0,0 +1,411 @@
+#include "cryptography/blob.hpp"
+#include "cryptography/crypto_provider/crypto_defaults.hpp"
+#include "framework/specified_visitor.hpp"
+#include "module/shared_model/builders/protobuf/test_block_builder.hpp"
+#include "module/shared_model/builders/protobuf/test_transaction_builder.hpp"
+#include "validators/field_validator.hpp"
+
+using namespace shared_model::proto;
+using namespace iroha::expected;
+using namespace shared_model::interface::types;
+
+using framework::SpecifiedVisitor;
+using shared_model::crypto::Blob;
+using shared_model::validation::FieldValidator;
+
+class ProtoQueryResponseFactoryTest : public ::testing::Test {
+ public:
+  std::shared_ptr<ProtoQueryResponseFactory> response_factory =
+      std::make_shared<ProtoQueryResponseFactory>();
+  std::shared_ptr<ProtoCommonObjectsFactory<FieldValidator>> objects_factory =
+      std::make_shared<ProtoCommonObjectsFactory<FieldValidator>>();
+
+  /**
+   * Put value of Result<unique_ptr<_>, _> into a shared_ptr
+   * @tparam ResultType - type of result value inside a unique_ptr
+   * @tparam ErrorType - type of result error
+   * @param res - result to be unwrapped
+   * @return shared_ptr to result value or to nullptr, if result containts error
+   */
+  template <typename ResultType, typename ErrorType>
+  std::unique_ptr<ResultType> unwrapResult(
+      Result<std::unique_ptr<ResultType>, ErrorType> &&res) {
+    return res.match(
+        [](Value<std::unique_ptr<ResultType>> &val) {
+          return std::move(val.value);
+        },
+        [](const Error<ErrorType> &) -> std::unique_ptr<ResultType> {
+          return nullptr;
+        });
+  }
+
+  void SetUp() override {}
+  void TearDown() override {}
+};
+
+/**
+ * Checks createAccountAssetResponse method of QueryResponseFactory
+ * @given collection of account assets
+ * @when creating account asset query response via factory
+ * @then that response is created @and is well-formed
+ */
+TEST_F(ProtoQueryResponseFactoryTest, CreateAccountAssetResponse) {
+  const HashType kQueryHash{"my_super_hash"};
+
+  constexpr int kAccountAssetsNumber = 5;
+  const std::string kAccountId = "doge@meme";
+  const std::string kAssetId = "dogecoin#iroha";
+
+  std::vector<std::shared_ptr<shared_model::interface::AccountAsset>> assets;
+  for (auto i = 1; i < kAccountAssetsNumber; ++i) {
+    auto asset = unwrapResult(objects_factory->createAccountAsset(
+        kAccountId,
+        kAssetId,
+        shared_model::interface::Amount(std::to_string(i))));
+    if (not asset) {
+      FAIL() << "could not create common object via factory";
+    }
+    assets.push_back(std::move(asset));
+  }
+  auto query_response =
+      response_factory->createAccountAssetResponse(assets, kQueryHash);
+
+  ASSERT_TRUE(query_response);
+  ASSERT_EQ(query_response->queryHash(), kQueryHash);
+  ASSERT_NO_THROW({
+    const auto &response = boost::apply_visitor(
+        SpecifiedVisitor<shared_model::interface::AccountAssetResponse>(),
```

Comment:
SpecifiedVisitor<T> can be replaced with boost::get<const T&> (const & in case of interface), since both return value or throw an exception.

Reply:
A scoped trace will not be required if boost::get is used.